cbits/cbits.c

```diff
@@ -0,0 +1,56 @@
+#include <stdint.h>
+#include <string.h>
+#include <limits.h>
+#include <assert.h>
+
+/* Taken from:
+ *
+ * <http://stackoverflow.com/questions/776508/best-practices-for-circular-shift-rotate-operations-in-c>
+ *
+ * Intel is little-endian, we have to rotate right
+ */
+static inline uint32_t rotr32(uint32_t n, unsigned int c) {
+  const unsigned int mask = (CHAR_BIT*sizeof(n)-1);
+  c &= mask;  // avoid undef behaviour with NDEBUG.  0 overhead for most types / compilers
+  return (n>>c) | (n<<( (-c)&mask ));
+}
+
+/* - mask_shift is the initial shift of the mask.  It is specified in bytes.
+ */
+void _hs_mask_chunk(
+        uint32_t mask, int mask_shift,
+        uint8_t *payload_start, size_t payload_len,
+        uint8_t *target) {
+    const uint8_t *payload_end = payload_start + payload_len;
+
+    uint8_t *p = payload_start;
```

Comment:
I agree, that's why I did the unaligned tests - and there is no noticable difference. This is allowed only on X86 architectures that do allow it, so it will not result in any segfault. I would leave it unaligned (maybe with a comment). I'd guess, that ultimately the unaligned-fast path will not be slower than the slow-path on modern architectures.